cpp/src/arrow/tensor.h

```diff
@@ -55,8 +56,76 @@ static inline bool is_tensor_supported(Type::type type_id) {
 template <typename SparseIndexType>
 class SparseTensorImpl;
 
+namespace internal {
+
+ARROW_EXPORT
+bool IsTensorStridesContiguous(const std::shared_ptr<DataType>& type,
+                               const std::vector<int64_t>& shape,
+                               const std::vector<int64_t>& strides);
+
+ARROW_EXPORT
+Status ValidateTensorParameters(const std::shared_ptr<DataType>& type,
+                                const std::shared_ptr<Buffer>& data,
+                                const std::vector<int64_t>& shape,
+                                const std::vector<int64_t>& strides,
+                                const std::vector<std::string>& dim_names);
+
+}  // namespace internal
+
 class ARROW_EXPORT Tensor {
  public:
+  /// \brief Create a Tensor with full parameters
+  ///
+  /// This factory function will return Status::Invalid when the parameters are
+  /// inconsistent
+  ///
+  /// \param[in] type The data type of the tensor values
+  /// \param[in] data The buffer of the tensor content
+  /// \param[in] shape The shape of the tensor
+  /// \param[in] strides The strides of the tensor
+  ///            (if this is empty, the data assumed to be row-major)
+  /// \param[in] dim_names The names of the tensor dimensions
+  static Result<std::shared_ptr<Tensor>> Make(const std::shared_ptr<DataType>& type,
+                                              const std::shared_ptr<Buffer>& data,
+                                              const std::vector<int64_t>& shape,
+                                              const std::vector<int64_t>& strides,
+                                              const std::vector<std::string>& dim_names) {
```

Comment:
I think (but am not sure) you can write `..., const std::vector<int64_t>& strides = {}, const std::vector<std::string>& dim_names = {})`